### PR TITLE
compiler: lazily generate ServiceDescriptors, and define equality

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
This does both changes discussed.

### CODE REVIEW

File: core/src/main/java/io/grpc/ServiceDescriptor.java [hunk 1]
```diff
@@ -67,4 +67,19 @@ public String getName() {
...
   }
+
+  @Override
+  public boolean equals(Object other) {
+    if (!(other instanceof ServiceDescriptor)) {
+      return false;
+    }
+    ServiceDescriptor that = (ServiceDescriptor) other;
+    return name.equals(that.name)
+        && methods.equals(that.methods);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ServiceDescriptor.java]: Method order should not technically impact equals. But it's hard to know how much that matters without having a use-case.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ServiceDescriptor.java]: Normalized.  

It seems ServiceDescriptor should probably have been more stringent, else equality would have been easier to define.

File: core/src/main/java/io/grpc/ServiceDescriptor.java [hunk 2]
```diff
@@ -67,4 +69,39 @@ public String getName() {
...
+      @Override
+      public int compare(MethodDescriptor<?, ?> o1, MethodDescriptor<?, ?> o2) {
+        // MethodDescriptors are currently reference-equal only.  This also happens to handle the
+        // case where there are duplicate MethodDescriptors present.
+        return System.identityHashCode(o1) - System.identityHashCode(o2);
+      }
+    };
+    Collections.sort(thisMethodsCopy, compare);
+    Collections.sort(thatMethodsCopy, compare);
+    return thisMethodsCopy.equals(thatMethodsCopy);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ServiceDescriptor.java]: This doesn't work if there are hash collisions. Just throw them into a set and see if they are equal?

File: core/src/main/java/io/grpc/ServiceDescriptor.java [hunk 3]
```diff
@@ -67,4 +69,39 @@ public String getName() {
...
+    // Super slow path, the price of doing tricky things.  Don't be tricky.
+    List<MethodDescriptor<?, ?>> thisMethodsCopy = new ArrayList<MethodDescriptor<?, ?>>(methods);
+    List<MethodDescriptor<?, ?>> thatMethodsCopy =
+        new ArrayList<MethodDescriptor<?, ?>>(that.methods);
+    Comparator<MethodDescriptor<?, ?>> compare = new Comparator<MethodDescriptor<?, ?>>() {
+      @Override
+      public int compare(MethodDescriptor<?, ?> o1, MethodDescriptor<?, ?> o2) {
+        // MethodDescriptors are currently reference-equal only.  This also happens to handle the
+        // case where there are duplicate MethodDescriptors present.
+        return System.identityHashCode(o1) - System.identityHashCode(o2);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ServiceDescriptor.java]: This doesn't correctly deal with underflow, which is very possible with uniformly distributed integers.

File: core/src/main/java/io/grpc/ServiceDescriptor.java (hunk 2 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ServiceDescriptor.java]: Wouldn't work if there were duplicate MethodsDescriptors.

Comment (ejona86) [inline core/src/main/java/io/grpc/ServiceDescriptor.java]: If there are true duplicates, then it seems safe to throw them away. A much bigger problem would be if there are two _different_ method descriptors for the same method name, but things would go badly with that anyway.

Wow, there really isn't any validation, including no verification that the methods are in the service. We should probably move or copy the validation in ServerServiceDefinition here.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ServiceDescriptor.java]: Wouldn't that be API breaking?

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ServiceDescriptor.java]: I actually don't care that much about defining equality.  I'm going to back out this file from the PR.

**Pull request merged**